packages/core/src/render3/i18n/i18n_hydration.ts

```diff
@@ -0,0 +1,252 @@
+
+export function setIsI18nHydrationSupportEnabled(enabled: boolean) {
+  _isI18nHydrationSupportEnabled = enabled;
+}
+
+export function isI18nHydrationSupportEnabled() {
+  return _isI18nHydrationSupportEnabled;
+}
+
+function trySerializeI18nBlockImpl(
+    lView: LView, index: number, processTextNode: (node: RNode) => void): Array<number>|null {
+  if (!isI18nHydrationSupportEnabled()) {
+    return null;
+  }
+
+  const tView = lView[TVIEW];
+  const tI18n = tView.data[index] as TI18n | undefined;
+
+  if (!tI18n || !tI18n.ast) {
+    return null;
+  }
+
+  const icuCases: number[] = [];
+  function serializeI18nBlock(node: I18nNode) {
+    switch (node.kind) {
+      case I18nNodeKind.TEXT:
+        const rNode = unwrapRNode(lView[node.index]!);
+        processTextNode(rNode);
+        break;
+
+      case I18nNodeKind.ELEMENT:
+      case I18nNodeKind.PLACEHOLDER:
+        node.children.forEach(serializeI18nBlock);
+        break;
+
+      case I18nNodeKind.ICU:
+        const currentCase = lView[node.currentCaseLViewIndex] as number;
+
+        if (currentCase < 0) {
+          const caseIdx = ~currentCase;
+          icuCases.push(caseIdx);
+          node.cases[caseIdx].forEach(serializeI18nBlock);
+        }
+        break;
+    }
+  }
+
+  tI18n.ast.forEach(serializeI18nBlock);
+  return icuCases.length > 0 ? icuCases : null;
+}
+
+let _trySerializeI18nBlockImpl: typeof trySerializeI18nBlockImpl = () => {
+  // noop unless `enableTrySerializeI18nBlock` is invoked.
+  return null;
+};
+
+export function enableTrySerializeI18nBlockImpl() {
+  _trySerializeI18nBlockImpl = trySerializeI18nBlockImpl;
+}
+
+/**
+ * Attempts to serialize i18n data for the given i18n block.
+ *
+ * @param lView lView with the i18n block
+ * @param index index of the i18n block in the lView
+ * @param processTextNode callback run for text nodes
+ * @returns the i18n data, or null if there is no relevant data
+ */
+export function trySerializeI18nBlock(
+    lView: LView, index: number, processTextNode: (node: RNode) => void): Array<number>|null {
+  return _trySerializeI18nBlockImpl(lView, index, processTextNode);
+}
+
+function prepareI18nBlockForHydrationImpl(lView: LView, index: number) {
+  if (!isI18nHydrationSupportEnabled()) {
+    return;
+  }
+
+  const hydrationInfo = lView[HYDRATION];
+  if (!hydrationInfo) {
+    return;
+  }
+
+  const tView = lView[TVIEW];
+  const tI18n = tView.data[index] as TI18n;
+
+  ngDevMode && assertDefined(tI18n, 'Expected i18n data');
+
+  const firstNode = tI18n.ast[0];
+  if (!firstNode) {
+    return;
+  }
+
+  const tNode = tView.data[firstNode.index] as TNode;
+  const rootNode: Node = locateNextRNode(hydrationInfo, tView, lView, tNode) as Node;
+
+  ngDevMode && assertDefined(rootNode, 'expected root node');
+
+  const disconnectedNodes = hydrationInfo.disconnectedNodes ??=
+      new Set(hydrationInfo.data[DISCONNECTED_NODES] ?? []);
+  const i18nMap: Map<number, Node> = hydrationInfo.i18nNodes ??= new Map();
```

Comment:
This might be sufficient to set the right type of the `i18nMap` const.
```suggestion
  const i18nMap = hydrationInfo.i18nNodes ??= new Map<number, Node>();
```